language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+
+```ruby
+Puppet::Type.newtype(:service) do
+  feature :enableable
+end
+```
+
+And the type specifies that the `enable` property can only be managed if the selected provider supports the `enableable` feature. This validation check is performed at catalog application time, once the provider has been resolved on the agent.
+
+```ruby
+Puppet::Type.newtype(:service) do
+  newproperty(:enable, :required_features => :enableable) do
+  end
+end
+```
+
+A **provider** indicates it supports the feature using `has_feature`:
+
+```ruby
+Puppet::Type.type(:service).provide :launchd, :parent => :base do
+  has_feature :enableable
+end
+```
+
+A type can also restrict allowed parameter/property values based on provider features. For example, the `enable` property can only be set to `mask` if the provider is `maskable`:
+
+```ruby
+Puppet::Type.newtype(:service) do
+  feature :maskable
+
+  newproperty(:enable, :required_features => :enableable) do
+    ...
+    newvalue(:mask, :event => :service_disabled, :required_features => :maskable) do
+      provider.mask
+    end
+  end
+end
+```
+
+### mixins
+
+Since puppet resource types are defined in ruby, you can mixin additional functionality. The AWS module uses this to create subclasses for different types of Route53 DNS records:
+
+```ruby
+Puppet::Type.newtype(:route53_a_record) do
+  extend PuppetX::Puppetlabs::Route53Record
+  @doc = 'Type representing a Route53 DNS record.'
+  create_properties_and_params()
+end
+```
+
+where the base module defines common parameters and properties for all `Route53Records`:
+
+```ruby
+module PuppetX
+  module Puppetlabs
+    module Route53Record
+      def create_properties_and_params
+        ensurable
+        newproperty(:zone) do
+        end
+      end
+    end
+  end
+end
+```
+
+## Parameter & Property DSL Methods
+
+The following DSL methods are commonly used for individual parameters and properties within the body of `newparam` and `newproperty` respectively:
+
+### desc
+
+Description of the property, output when running `puppet describe <type>`:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:region) do
+    desc 'The region in which to launch the instance.'
+  end
+end
+```
+
+### isnamevar
+
+Specifies that the parameter is the `namevar` (aka identity) for the resource:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:name) do
+    isnamevar
+  end
+end
+```
+
+As mentioned earlier, if a type specifies a parameter named `:name`, it will automatically be the namevar, so the call to `isnamevar` is redundant, but is explicit.
+
+Alternatively, you can pass an option when calling newparam:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:name, :namevar => true) do
+  end
+end
+```
+
+Note that the namevar is necessarily a parameter, and not a property, since changing the name identifies a different resource as opposed to changing the name of an existing resource.
```

Comment:
It may be worth calling out [package's use](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/package.rb#L268-L272) of namevar. We had a case there that we want to make the provider name part of the namevar. Future changes to the Type API will likely need some way to support this.